### ISSUE TITLE
hfst-proc: don't always uppercase everything during generation

### ISSUE THREAD

**Issue by unhammer** [hfst/hfst#320]
hfst-proc, imitating lt-proc, allows generating JA from '^JA<cnjcoo>$' even though the FST only has ja<cnjcoo>:ja (here optimised-lookup gives unknown).

But this "overgenerates" with words that have mixed case, e.g. acronyms with case endings. We want NATO:n from '^NATO<n><acr><sg><ine>$', but get NATO:N (whereas hfst-optimised-lookup gives the correct NATO:n). 

Is there a good solution to this? (what does lt-proc do?)

**Comment by unhammer** [comment]
```
<dictionary>
<sdefs>
<sdef n="gen"/>
<sdef n="nom"/>
</sdefs>
<section id="main" type="standard">
<e><p><l>cake</l><r>cake<s n="nom"/></r></p></e>
<e><p><l>cake:s</l><r>cake<s n="gen"/></r></p></e>
<e><p><l>NATO</l><r>NATO<s n="nom"/></r></p></e>
<e><p><l>NATO:s</l><r>NATO<s n="gen"/></r></p></e>
</section>
</dictionary>
```

```
$ lt-comp rl foo.dix foo.bin
main@standard 7 7

$ echo '^NATO<gen>$ ^CAKE<gen>$'|lt-proc -g foo.bin
NATO:s CAKE:S

$ lt-print foo.bin|sed 's/ /@_SPACE_@/g'| hfst-txt2fst  -e ε|hfst-fst2fst -O >foo.hfstol

$ echo '^NATO<gen>$ ^CAKE<gen>$'|hfst-proc -g foo.hfstol
NATO:S CAKE:S


```

**Comment by unhammer** [comment]
hfst-proc/applicators.cc says
```
bool
GenerationApplicator::lookup(const TokenVector& tokens, bool generate_on_fail)
{
  LookupState state(transducer);
  state.lookup(token_stream.to_symbols(tokens), caps_mode);
  if(state.is_final()) // generation succeeded
  {
    CapitalizationState capitalization_state = 
      caps_mode==DictionaryCase ?
        Unknown :
        token_stream.get_capitalization_state(TokenVector(tokens.begin(),tokens.size()>1?tokens.begin()+2:tokens.end()));
```
– so the caps-state seems to be solely based on the input. 

Then that caps-state is used to do
`    token_stream.put_symbols((*finals.begin())->get_output_symbols(),capitalization_state);`
which through symbols_to_string just uppercases based on the caps-state, without anywhere taking the dictionary-case into account.

**Comment by unhammer** [comment]
closed in bf26db9cfb09c5502e04c2c49253dd0a7231f384
